aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move

```diff
@@ -1601,10 +1586,237 @@ module supra_framework::pbo_delegation_pool {
+        // short-circuit if amount to reactivate is 0 so no event is emitted
+        if (amount == 0) { return };
+        // fail unlock of less than `MIN_COINS_ON_SHARES_POOL`
+        assert!(amount >= MIN_COINS_ON_SHARES_POOL, error::invalid_argument(EMINIMUM_UNLOCK_AMOUNT));
+        // synchronize delegation and stake pools before any user operation
+        synchronize_delegation_pool(pool_address);
+
+        let pool = borrow_global_mut<DelegationPool>(pool_address);
+
+        amount = coins_to_transfer_to_ensure_min_stake(
+            pending_inactive_shares_pool(pool),
+            &pool.active_shares,
+            delegator,
+            amount
+        );
+        let observed_lockup_cycle = pool.observed_lockup_cycle;
+        amount = redeem_inactive_shares(
+            pool,
+            delegator,
+            amount,
+            observed_lockup_cycle
+        );
+
+        stake::reactivate_stake(&retrieve_stake_pool_owner(pool), amount);
+
+        buy_in_active_shares(pool, delegator, amount);
+        assert_min_active_balance(pool, delegator);
+
+        event::emit_event(
+            &mut pool.reactivate_stake_events,
+            ReactivateStakeEvent {
+                pool_address,
+                delegator_address: delegator,
+                amount_reactivated: amount
+            }
+        );
+    }
+
+    /// For each delegator in `delegators`, locks the amount of stake specified in the same index of `stakes`.
+    /// The locked amount is subject to the vesting schedule specified when the delegation pool corresponding
+    /// to `pool_address` was created.
+    /// 
+    /// This function does not actually add the `stakes` to the stake pool, so must remain private and must
+    /// only be called either after ensuring that all `stakes` already exist in the `active` stake pool, or
+    /// will do so before the end of the transaction.
+    /// 
+    /// This function also assumes that `multisig_admin` is the admin of the `DelegationPool` at `pool_address`.
+    /// The caller must ensure that this is true.
+    fun lock_existing_delegators_stakes(
+        pool_address: address,
+        delegators: vector<address>,
+        stakes: vector<u64>
+    ) acquires DelegationPool {
+        let principle_stake_table =
+            &mut (borrow_global_mut<DelegationPool>(pool_address).principle_stake);
+
+        vector::zip_reverse(
+            delegators,
+            stakes,
+            |delegator, stake| {
+                if (stake > 0) {
+                    // Compute the stake to be added. `principle_stake` must be added to the table accordingly.
+                    if (table::contains(principle_stake_table, delegator)) {
+                        let stake_amount =
+                            table::borrow_mut(principle_stake_table, delegator);
+                        *stake_amount = *stake_amount + stake;
```

Comment:
```suggestion
                        *stake_amount = stake;
```
I think it would be more general if we just set it to the new value rather than add to the existing value.

Reply:
No, it won't be because `fund_delegators_with_locked_stake` is intended to be additive.

Reply:
I'll make a new version of this function to call from `lock_delegators_stakes`. I agree that it would be better for us to be able to decrease the amount of locked stake if necessary.